File: Lowest_Common_ancestor_in_Binary_Tree.cpp
//Time Complexity-O(n)-->'n' is number of nodes
//Space Complexity-O(1) if recursive stack space is not considered or else it is O(h)-->'h' is height of Binary tree
//Did the code run on Leetcode? Yes

class Solution {
public:
    TreeNode* res=NULL;
    bool lowest(TreeNode* root,TreeNode* p,TreeNode* q)
    {
        if(root==NULL)
        {
            return 0;
        }
        int mid=(root->val==p->val|| root->val==q->val)?1:0;
        int left=lowest(root->left,p,q);
        int right=lowest(root->right,p,q);
        if( mid+left+right>=2)
        {
            res=root;
        }
        return mid+left+right>0;
        
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        lowest(root,p,q);
        return res;
    }
};

File: lowestCommonAncestorInBinaryTree.java
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

/**
 * 
 */

public class lowestCommonAncestorInBinaryTree {
    TreeNode result;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {        
        traverse(root,p,q);
        return result;
    }

    private boolean traverse(TreeNode root, TreeNode p, TreeNode q){
        if(root == null){
            return false; 
        }

        boolean left = traverse(root.left,p,q);
        boolean right = traverse(root.right,p,q);

        if(left && right){ 
            result = root;
        }

        if(root.equals(p) || root.equals(q)){
            if(left || right){ 
                result = root;
                return false;
            }
            return true; 
        }

        return left || right;
    } 
}
